### PR TITLE
Feature/rchain persist processing step scripts

### LINKED ISSUE #670: rChain: persist calculation scripts
- [x] processing_step_calculation table: add column script type text
- [x] RChainService.persistUserScripts - persist calculation scripts 
- [x] rChain initialization: populate every user/step-nnn/[name].R with content of corresponding processing_step_calculation table script column 
- [x] processingChainApi/Repository: make sure when fetching processingStepCalculation(s) the column script is not included

### PR DESCRIPTION
**Pull request by SteRiccio**
Fixes #670

### CODE REVIEW

File: server/modules/analysis/service/rChainService/index.js [hunk 1]
```diff
@@ -6,14 +6,17 @@ import * as CSVReader from '@server/utils/file/csvReader'
...
 import * as NodeDef from '@core/survey/nodeDef'
 import * as ProcessingStep from '@common/analysis/processingStep'
+import * as ProcessingStepCalculation from '@common/analysis/processingStepCalculation'
 import * as ResultStepView from '@common/surveyRdb/resultStepView'
 
 import * as SurveyManager from '@server/modules/survey/manager/surveyManager'
 import * as SurveyRdbMamager from '@server/modules/surveyRdb/manager/surveyRdbManager'
 
 import * as ProcessingChainManager from '../../manager/processingChainManager'
+import * as CalculationManager from '../../manager/calculation'
```
Comment (minotogna) [inline server/modules/analysis/service/rChainService/index.js:16]: CalculationManager must not exist. Only AnalysisManager

File: server/modules/analysis/service/rChainService/index.js [hunk 2]
```diff
@@ -59,11 +62,34 @@ export const persistResults = async (surveyId, cycle, stepUuid, filePath) => {
...
+    entryNames.find((entryName) => new RegExp(`^${folder}\\/\\d{3}-${name}\\.R$`).test(entryName))
 
   await db.tx(async (tx) => {
-    const scriptCommon = await fileZip.getEntryAsText(findEntry('common'))
+    // Persist common script
+    const scriptCommon = await fileZip.getEntryAsText(findEntry(RChain.dirNames.user, 'common'))
     await ProcessingChainManager.updateChainScriptCommon(surveyId, chainUuid, scriptCommon, tx)
+
+    // Persist calculation scripts
+    const steps = await ProcessingChainManager.fetchStepsAndCalculationsByChainUuid(surveyId, chainUuid, tx)
```
Comment (minotogna) [inline server/modules/analysis/service/rChainService/index.js:77]: use ```AnalysisManager.fetchChain({chainUuid, includeStepsAndCalculations: true})```

File: server/modules/analysis/service/rChainService/index.js [hunk 3]
```diff
@@ -59,11 +62,34 @@ export const persistResults = async (surveyId, cycle, stepUuid, filePath) => {
...
-  const findEntry = (name) => entryNames.find((entryName) => !!entryName.match(new RegExp(`\\d{3}-${name}\\.R`)))
+
+  const findEntry = (folder, name) =>
+    entryNames.find((entryName) => new RegExp(`^${folder}\\/\\d{3}-${name}\\.R$`).test(entryName))
 
   await db.tx(async (tx) => {
-    const scriptCommon = await fileZip.getEntryAsText(findEntry('common'))
+    // Persist common script
+    const scriptCommon = await fileZip.getEntryAsText(findEntry(RChain.dirNames.user, 'common'))
     await ProcessingChainManager.updateChainScriptCommon(surveyId, chainUuid, scriptCommon, tx)
```
Comment (minotogna) [inline server/modules/analysis/service/rChainService/index.js:74]: Refactor this with AnalysisManager

File: server/modules/analysis/service/rChainService/index.js [hunk 4]
```diff
@@ -59,11 +62,34 @@ export const persistResults = async (surveyId, cycle, stepUuid, filePath) => {
...
+      steps.map((step) => {
+        const stepFolder = `${RChain.dirNames.user}/${RStep.getSubFolder(step)}`
+        return Promise.all(
+          ProcessingStep.getCalculations(step).map(async (calculation) => {
+            // Persist the script of each calculation
+            const calculationUuid = ProcessingStepCalculation.getUuid(calculation)
+            const nodeDefUuid = ProcessingStepCalculation.getNodeDefUuid(calculation)
+            const nodeDefName = NodeDef.getName(Survey.getNodeDefByUuid(nodeDefUuid)(survey))
+            const script = await fileZip.getEntryAsText(findEntry(stepFolder, nodeDefName))
+            return CalculationManager.updateCalculationScript(surveyId, calculationUuid, script, tx)
```
Comment (minotogna) [inline server/modules/analysis/service/rChainService/index.js:89]: use AnalysisManager

File: server/modules/analysis/repository/calculation/update.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import { TableCalculation } from '@common/model/db'
+
+import { db } from '@server/db/db'
```
Comment (minotogna) [inline server/modules/analysis/repository/calculation/update.js:3]: let's start using relative imports in server/common/core

File: server/modules/analysis/service/rChainService/rCalculation.js [hunk 6]
```diff
@@ -47,7 +50,10 @@ export default class RCalculation {
         NodeDef.getName
       )(this.calculation)
 
-      await this.rFile.appendContent(setVar(dfVar(entityName, attributeName), NA))
+      const calculationUuid = ProcessingStepCalculation.getUuid(this.calculation)
+      const script = await CalculationManager.fetchCalculationScript({ surveyId, calculationUuid })
```
Comment (minotogna) [inline server/modules/analysis/service/rChainService/rCalculation.js:54]: CalculationManager should not exist, neither the method CalculationManager.fetchCalculationScript.
In rChain init, use fetchChain with includeStript = true and you already have the script associated to the calculation

File: server/modules/analysis/service/rChainService/index.js (hunk 1 above)
Comment (SteRiccio) [reply server/modules/analysis/service/rChainService/index.js:16]: ok

File: server/modules/analysis/service/rChainService/index.js (hunk 2 above)
Comment (SteRiccio) [reply server/modules/analysis/service/rChainService/index.js:77]: ok

File: server/modules/analysis/service/rChainService/index.js (hunk 3 above)
Comment (SteRiccio) [reply server/modules/analysis/service/rChainService/index.js:74]: ok

File: server/modules/analysis/service/rChainService/index.js (hunk 4 above)
Comment (SteRiccio) [reply server/modules/analysis/service/rChainService/index.js:89]: ok

File: server/modules/analysis/repository/calculation/update.js (hunk 5 above)
Comment (SteRiccio) [reply server/modules/analysis/repository/calculation/update.js:3]: ok

File: server/modules/analysis/service/rChainService/rCalculation.js (hunk 6 above)
Comment (SteRiccio) [reply server/modules/analysis/service/rChainService/rCalculation.js:54]: ok, great

File: server/modules/analysis/repository/chain/read.js [hunk 7]
```diff
@@ -38,3 +38,27 @@ export const fetchChains = async (params, client = db) => {
...
+ * @param {pgPromise.IDatabase} [client=db] - The database client.
+ *
+ * @returns {Promise<any[]>} - The result promise.
+ */
+export const fetchChainByUuid = async (params, client = db) => {
+  const { surveyId, chainUuid, includeScript = false, includeStepsAndCalculations = false } = params
+
+  const tableChain = new TableChain(surveyId)
+
+  return client.one(
```
Comment (minotogna) [inline server/modules/analysis/repository/chain/read.js:59]: should we use ```oneOrNone```?

File: server/modules/analysis/repository/chain/read.js [hunk 8]
```diff
@@ -38,3 +38,27 @@ export const fetchChains = async (params, client = db) => {
...
+ * Fetches a single processing chain by the given survey id and chainUuid.
+ *
+ * @param {!object} params - The query parameters.
+ * @param {!string} params.surveyId - The survey id.
+ * @param {!string} params.chainUuid - The processing chain uuid.
+ * @param {boolean} [params.includeStepsAndCalculations=false] - Whether to include the processing steps and calculations.
+ * @param {boolean} [params.includeScript=false] - Whether to include the R scripts.
+ * @param {pgPromise.IDatabase} [client=db] - The database client.
+ *
+ * @returns {Promise<any[]>} - The result promise.
```
Comment (minotogna) [inline server/modules/analysis/repository/chain/read.js:52]: return type is ProcessingChain or null (if we use oneOrNone) ```@returns {Promise<ProcessingChain|null>} - The result promise.```

File: server/modules/analysis/repository/chain/update.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+ * Updates the common script of the given processing chain.
+ *
+ * @param {!string} surveyId - The survey id.
+ * @param {!string} chainUuid - The processing chain uuid.
+ * @param {!string} script - The script content.
+ * @param {pgPromise.IDatabase} [client=db] - The database client.
+ *
+ * @returns {Promise<any[]>} - The result promise.
+ */
+export const updateChainScriptCommon = (surveyId, chainUuid, script, client = db) => {
```
Comment (minotogna) [inline server/modules/analysis/repository/chain/update.js:17]: use params as first argument

File: server/modules/analysis/repository/chain/update.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * Updates the common script of the given processing chain.
+ *
+ * @param {!string} surveyId - The survey id.
+ * @param {!string} chainUuid - The processing chain uuid.
+ * @param {!string} script - The script content.
+ * @param {pgPromise.IDatabase} [client=db] - The database client.
+ *
+ * @returns {Promise<any[]>} - The result promise.
```
Comment (minotogna) [inline server/modules/analysis/repository/chain/update.js:15]: use params and params.surveyId ... 
return ```Promise<null>```

File: server/modules/analysis/repository/chain/read.js [hunk 11]
```diff
@@ -38,3 +38,27 @@ export const fetchChains = async (params, client = db) => {
...
+ */
+export const fetchChainByUuid = async (params, client = db) => {
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply server/modules/analysis/repository/chain/read.js:54]: ok

File: server/modules/analysis/repository/chain/read.js (hunk 7 above)
Comment (SteRiccio) [reply server/modules/analysis/repository/chain/read.js:59]: right

File: server/modules/analysis/repository/chain/read.js (hunk 8 above)
Comment (SteRiccio) [reply server/modules/analysis/repository/chain/read.js:52]: right

File: server/modules/analysis/repository/chain/update.js (hunk 9 above)
Comment (SteRiccio) [reply server/modules/analysis/repository/chain/update.js:17]: ok

File: server/modules/analysis/repository/chain/update.js (hunk 10 above)
Comment (SteRiccio) [reply server/modules/analysis/repository/chain/update.js:15]: sure

File: server/modules/analysis/service/rChainService/index.js [hunk 12]
```diff
@@ -56,11 +58,37 @@ export const persistResults = async (surveyId, cycle, stepUuid, filePath) => {
...
+    // Persist common script
+    const scriptCommon = await fileZip.getEntryAsText(findEntry(RChain.dirNames.user, 'common'))
+    await AnalysisManager.updateChainScriptCommon({ surveyId, chainUuid, scriptCommon }, tx)
+
+    // Persist calculation scripts
+    const chain = await AnalysisManager.fetchChain(
+      { surveyId, chainUuid, includeScript: true, includeStepsAndCalculations: true },
+      tx
+    )
+    const survey = await SurveyManager.fetchSurveyAndNodeDefsBySurveyId(surveyId)
```
Comment (minotogna) [inline server/modules/analysis/service/rChainService/index.js:77]: what about Promise.all?

Comment (SteRiccio) [reply server/modules/analysis/service/rChainService/index.js:77]: right, it's better

**Pull request merged**